src/meta/processors/admin/CreateBackupProcessor.cpp

```diff
@@ -0,0 +1,277 @@
+
+    return files;
+}
+
+folly::Optional<std::unordered_set<GraphSpaceID>> CreateBackupProcessor::spaceNameToId(
+    const std::vector<std::string>* backupSpaces) {
+    std::unordered_set<GraphSpaceID> spaces;
+
+    if (backupSpaces != nullptr) {
+        DCHECK(!backupSpaces->empty());
+        std::vector<std::string> values;
+        std::vector<std::string> keys;
+
+        values.reserve(backupSpaces->size());
+        keys.reserve(backupSpaces->size());
+
+        std::transform(
+            backupSpaces->begin(), backupSpaces->end(), std::back_inserter(keys), [](auto& name) {
+                return MetaServiceUtils::indexSpaceKey(name);
+            });
+
+        auto ret = kvstore_->multiGet(kDefaultSpaceId, kDefaultPartId, std::move(keys), &values);
+        if (ret.first != kvstore::ResultCode::SUCCEEDED) {
+            LOG(ERROR) << "Failed to get space id, error: " << ret.first;
+            handleErrorCode(cpp2::ErrorCode::E_BACKUP_FAILURE);
+            return folly::none;
+        }
+
+        std::transform(std::make_move_iterator(values.begin()),
+                       std::make_move_iterator(values.end()),
+                       std::inserter(spaces, spaces.end()),
+                       [](const std::string& rawID) {
+                           return *reinterpret_cast<const GraphSpaceID*>(rawID.c_str());
+                       });
+
+    } else {
+        auto prefix = MetaServiceUtils::spacePrefix();
+        std::unique_ptr<kvstore::KVIterator> iter;
+        auto ret = kvstore_->prefix(kDefaultSpaceId, kDefaultPartId, prefix, &iter);
+        if (ret != kvstore::ResultCode::SUCCEEDED) {
+            handleErrorCode(MetaCommon::to(ret));
+            return folly::none;
+        }
+
+        while (iter->valid()) {
+            auto spaceId = MetaServiceUtils::spaceId(iter->key());
+            auto spaceName = MetaServiceUtils::spaceName(iter->val());
+            VLOG(3) << "List spaces " << spaceId << ", name " << spaceName;
+            spaces.emplace(spaceId);
+            iter->next();
+        }
+    }
+
+    if (spaces.empty()) {
+        LOG(ERROR) << "Failed to create a full backup because there is currently no space.";
+        handleErrorCode(cpp2::ErrorCode::E_BACKUP_SPACE_NOT_FOUND);
+        return folly::none;
+    }
+
+    return spaces;
+}
+
+void CreateBackupProcessor::process(const cpp2::CreateBackupReq& req) {
+    auto* backupSpaces = req.get_spaces();
+
+    auto result = MetaServiceUtils::isIndexRebuilding(kvstore_);
+    if (result == folly::none) {
+        handleErrorCode(cpp2::ErrorCode::E_BACKUP_FAILURE);
+        onFinished();
+        return;
+    }
+
+    if (!result.value()) {
+        handleErrorCode(cpp2::ErrorCode::E_BACKUP_BUILDING_INDEX);
+        onFinished();
+        return;
+    }
+
+    folly::SharedMutex::WriteHolder wHolder(LockUtils::snapshotLock());
+
+    auto hosts = ActiveHostsMan::getActiveHosts(kvstore_);
```

Comment:
Do you think we need to check all hosts online before we start? We didn't do it in 1.0

Reply:
I think it is needed because if some machine is offline at the time of the backup, then it is possible that the backup will fail.

Reply:
If so, this need to changed. `ActiveHostsMan::getActiveHosts` only get all active hosts, not the hosts contain specified spaces.